# Clinical Coding and Ontology Practical

First we will take a look at some ontology stuff. We will use some tools to take a look at different ontology related items.

In [83]:
# Set up Komenti
!rm -rf komenti
!rm komenti-0.2.0-SNAPSHOT.zip
!wget https://github.com/reality/komenti/releases/download/0.2.0-SNAPSHOT-4/komenti-0.2.0-SNAPSHOT.zip
!unzip komenti-0.2.0-SNAPSHOT.zip
!mv komenti-0.2.0-SNAPSHOT komenti

# and klarigi
!rm -rf klarigi
!wget http://lokero.xyz/klarigi-0.0.12-SNAPSHOT.zip
!unzip klarigi-0.0.12-SNAPSHOT.zip
!mv klarigi-0.0.12-SNAPSHOT klarigi

#and ontologies
!rm hp.owl
!wget http://purl.obolibrary.org/obo/hp.owl
!rm doid.owl
!wget http://purl.obolibrary.org/obo/doid.owl

rm: ne eblas forigi 'komenti-0.2.0-SNAPSHOT.zip': Dosiero aŭ dosierujo ne ekzistas
--2021-10-05 12:57:13--  https://github.com/reality/komenti/releases/download/0.2.0-SNAPSHOT-4/komenti-0.2.0-SNAPSHOT.zip
SSL_INIT
Ni ŝargis la CA-atestilon '/etc/ssl/certs/ca-certificates.crt'
Ni solvigas github.com (github.com)... 140.82.121.4
Konektado al github.com (github.com)|140.82.121.4|:443... konektita.
HTTP peto sendita, ni atendas respondon... 302 Found
Loko: https://github-releases.githubusercontent.com/225223734/d460ae80-a82e-11eb-910e-5f96167c2c3f?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211005%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211005T115713Z&X-Amz-Expires=300&X-Amz-Signature=f01dfd818b18c631d9759653daae4d9b5f21d49c9cb74cfe6018bef5a30b22d5&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=225223734&response-content-disposition=attachment%3B%20filename%3Dkomenti-0.2.0-SNAPSHOT.zip&response-content-type=application%2Foctet-stream [sekvanta]
--2

  inflating: komenti-0.2.0-SNAPSHOT/lib/trove4j-3.0.3.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/caffeine-2.5.6.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/guice-assistedinject-4.1.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/guice-multibindings-4.1.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/guice-4.1.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/guava-20.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/jsr305-3.0.1.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/commons-io-2.5.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/mapper-extras-client-7.12.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/parent-join-client-7.12.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/aggs-matrix-stats-client-7.12.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/rank-eval-client-7.12.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/lang-mustache-client-7.12.0.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/commons-codec-1.11.jar  
  inflating: komenti-0.2.0-SNAPSHOT/lib/ant-junit-1.9.13.jar  
  i

  inflating: klarigi-0.0.12-SNAPSHOT/lib/slib-graph-0.9.1.pom  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/slib-graph-model-impl-0.9.1.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/slib-graph-utils-0.9.1.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/slib-indexer-0.9.1.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/slib-graph-model-0.9.1.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/slib-utils-0.9.1.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-runtime-2.7.13.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-repository-manager-2.7.13.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-repository-http-2.7.13.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-sail-federation-2.7.13.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-sail-memory-2.7.13.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-sail-nativerdf-2.7.13.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-sail-rdbms-2.7.13.jar  
  inflating: klarigi-0.0.12-SNAPSHOT/lib/sesame-queryalgebra-evalua

--2021-10-05 12:58:18--  http://purl.obolibrary.org/obo/hp.owl
Ni solvigas purl.obolibrary.org (purl.obolibrary.org)... 52.3.123.63
Konektado al purl.obolibrary.org (purl.obolibrary.org)|52.3.123.63|:80... konektita.
HTTP peto sendita, ni atendas respondon... 302 Found
Loko: https://raw.githubusercontent.com/obophenotype/human-phenotype-ontology/master/hp.owl [sekvanta]
--2021-10-05 12:58:18--  https://raw.githubusercontent.com/obophenotype/human-phenotype-ontology/master/hp.owl
SSL_INIT
Ni ŝargis la CA-atestilon '/etc/ssl/certs/ca-certificates.crt'
Ni solvigas raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Konektado al raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... konektita.
HTTP peto sendita, ni atendas respondon... 200 OK
Grando: 85382374 (81M) [text/plain]
Ni konservas al: ‘hp.owl’

hp.owl              100%[===================>]  81,43M  8,05MB/s    en 10s     

2021-10-05 12:58:33 (7,9

# Part One: Loading and Exploring the EHR

Now we can start to take a look at our demo dataset, which in some way you should perhaps be familiar with. Let's download it:

In [6]:
!wget https://physionet.org/static/published-projects/mimiciii-demo/mimic-iii-clinical-database-demo-1.4.zip
!unzip mimic-iii-clinical-database-demo-1.4.zip
!mv mimic-iii-clinical-database-demo-1.4 mimic

--2021-10-03 15:36:58--  https://physionet.org/static/published-projects/mimiciii-demo/mimic-iii-clinical-database-demo-1.4.zip
SSL_INIT
Ni ŝargis la CA-atestilon '/etc/ssl/certs/ca-certificates.crt'
Ni solvigas physionet.org (physionet.org)... 18.18.42.54
Konektado al physionet.org (physionet.org)|18.18.42.54|:443... konektita.
HTTP peto sendita, ni atendas respondon... 200 OK
Grando: 11350216 (11M) [application/zip]
Ni konservas al: ‘mimic-iii-clinical-database-demo-1.4.zip’

mimic-iii-clinical- 100%[===================>]  10,82M  5,72MB/s    en 1,9s    

2021-10-03 15:37:01 (5,72 MB/s) - ‘mimic-iii-clinical-database-demo-1.4.zip’ konservita [11350216/11350216]

Archive:  mimic-iii-clinical-database-demo-1.4.zip
   creating: mimic-iii-clinical-database-demo-1.4/
  inflating: mimic-iii-clinical-database-demo-1.4/ADMISSIONS.csv  
  inflating: mimic-iii-clinical-database-demo-1.4/CALLOUT.csv  
  inflating: mimic-iii-clinical-database-demo-1.4/CAREGIVERS.csv  
  inflating: mimic-iii-clin

We can load some of these interesting tables into pandas, and take a look:

In [3]:
import pandas as pd

diagnoses = pd.read_csv("mimic/DIAGNOSES_ICD.csv")
diagnoses

row_id  subject_id  hadm_id  seq_num icd9_code
0     112344       10006   142345        1     99591
1     112345       10006   142345        2     99662
2     112346       10006   142345        3      5672
3     112347       10006   142345        4     40391
4     112348       10006   142345        5     42731
...      ...         ...      ...      ...       ...
1756  397673       44228   103379        7      1975
1757  397674       44228   103379        8     45182
1758  397675       44228   103379        9     99592
1759  397676       44228   103379       10      2449
1760  397677       44228   103379       11      2859

[1761 rows x 5 columns]

So, here we have a few columns:

* row_id: This identifies the row in this table (boring
* subject_id: Identifies the individual/patient
* hadm_id: This identifies the admission, or the particular stay
* seq_num: Identifies the order of the diagnosis
* icd9_code: An ICD-9 code referring to the diagnosis 

So, this table describes a list of diagnoses received for each patient admission, identified by ICD9. These diagnoses are applied by professional clinical coders, who apply these codes.  we also have definition information provided by another CSV file, which we can cross-reference with our diagnoses data frame to get some more information, or at least the names, of these diagnoses.

Let's first load the definitions table:

In [4]:
diag_defs = pd.read_csv("mimic/D_ICD_DIAGNOSES.csv")[['icd9_code', 'long_title']]
diag_defs

icd9_code                                         long_title
0         01716  Erythema nodosum with hypersensitivity reactio...
1         01720  Tuberculosis of peripheral lymph nodes, unspec...
2         01721  Tuberculosis of peripheral lymph nodes, bacter...
3         01722  Tuberculosis of peripheral lymph nodes, bacter...
4         01723  Tuberculosis of peripheral lymph nodes, tuberc...
...         ...                                                ...
14562     V8712   Contact with and (suspected) exposure to benzene
14563     V8719  Contact with and (suspected) exposure to other...
14564      V872  Contact with and (suspected) exposure to other...
14565     V8731      Contact with and (suspected) exposure to mold
14566     V8732  Contact with and (suspected) exposure to algae...

[14567 rows x 2 columns]

We can see here that we have label information for each ICD9 code. We can now do a simple join on the diagnoses table to add the disease titles to the table... 

In [5]:
diag_with_titles = diagnoses.merge(diag_defs, on="icd9_code", how="left")
diag_with_titles

row_id  subject_id  hadm_id  seq_num icd9_code  \
0     112344       10006   142345        1     99591   
1     112345       10006   142345        2     99662   
2     112346       10006   142345        3      5672   
3     112347       10006   142345        4     40391   
4     112348       10006   142345        5     42731   
...      ...         ...      ...      ...       ...   
1756  397673       44228   103379        7      1975   
1757  397674       44228   103379        8     45182   
1758  397675       44228   103379        9     99592   
1759  397676       44228   103379       10      2449   
1760  397677       44228   103379       11      2859   

                                             long_title  
0                                                Sepsis  
1     Infection and inflammatory reaction due to oth...  
2                                                   NaN  
3     Hypertensive chronic kidney disease, unspecifi...  
4                                   Atrial fibrillation  
...                                                 ...  
1756  Secondary malignant neoplasm of large intestin...  
1757  Phlebitis and thrombophlebitis of superficial ...  
1758                                      Severe sepsis  
1759                Unspecified acquired hypothyroidism  
1760                                Anemia, unspecified  

[1761 rows x 6 columns]

That will help us read it. This demonstrates the role of vocabulary codes to uniquely identify a particular concept: in this case a disease.

This is kind of nice, but one of the problems we have here is that ICD-9 is a serious outmoded. seriously I think it's older than me. This, and the lack of formal semantic features makes it difficult to do useful things with this terminology.

Well, we can use Komenti to download metadata assocatied with classes. However, that would take a little time, so let's take a look at once I prepared earlier 

# Part 2: Querying Biomedical Ontologies

So before we talked about the Disease Ontology (DO), and how this stratifies different diseases and so on. We can use a handy piece of software called Komenti to look at different diseases and so on.

To run a query on a biomedical ontology, we can use Komenti, which links to an ontology repository called AberOWL (which you can also browse on the web from https://aber-owl.net/)

In the following command, we use komenti to 

* -q: This argument identifies the query you're running. Note: if the label is more than one word, it needs to additionally have a single quote around it.
* --class-mode
* --query-type

You can get a full-ish list of the query arguments [here](https://github.com/reality/Komenti)


By changing the query type, we can also retrieve more specific kinds of disease:

In [2]:
!komenti query -q "cardiomyopathy" -o DOID --class-mode --query-type equivalent

SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
cardiomyopathy	http://purl.obolibrary.org/obo/DOID_0050700	cardiomyopathy	DOID	1


We can also retrieve, from the Human Phenotype ontology, different phenotypes:

In [ ]:
!komenti query -q "hypertension" -o HP --class-mode --query-type subeq

You can also retrieve classes directly from their IRIs. You just have to wrap the argument with angled brackets (< and >). Try going to [AberOWL](https://aber-owl.net/ontology/HP) and picking a phenotype you're interested in, then insert it into the query below:

In [ ]:
!komenti query -q "<[INSERT IRI HERE]>" -o HP --class-mode --query-type equivalent

# Part 3: Linking EHR Entities to Ontologies

Now, this can take a little bit of work to parse and match up with the . Especially if we were interested in all of them. Thankfully, we can take this handy DOID<->ICD9 mapping I created earlier!

In [6]:
!rm doid_icd_map*
!wget http://lokero.xyz/doid_icd_mappings.txt

doid_icd_map = pd.read_csv('doid_icd_mappings.txt', sep='\t')
doid_icd_map['icd9_code'] = doid_icd_map['icd9_code'].map(lambda icd: icd.replace('.', ''))

doid_icd_map

--2021-10-04 19:10:20--  http://lokero.xyz/doid_icd_mappings.txt
Ni solvigas lokero.xyz (lokero.xyz)... 176.58.107.169
Konektado al lokero.xyz (lokero.xyz)|176.58.107.169|:80... konektita.
HTTP peto sendita, ni atendas respondon... 200 OK
Grando: 101061 (99K) [text/plain]
Ni konservas al: ‘doid_icd_mappings.txt’

doid_icd_mappings.t 100%[===================>]  98,69K   386KB/s    en 0,3s    

2021-10-04 19:10:21 (386 KB/s) - ‘doid_icd_mappings.txt’ konservita [101061/101061]



icd9_code                                       doid
0        11510  http://purl.obolibrary.org/obo/DOID_11315
1         3310  http://purl.obolibrary.org/obo/DOID_10652
2         1150   http://purl.obolibrary.org/obo/DOID_1759
3        37205  http://purl.obolibrary.org/obo/DOID_11203
4        37945  http://purl.obolibrary.org/obo/DOID_14523
...        ...                                        ...
2113      1112  http://purl.obolibrary.org/obo/DOID_13902
2114     37333   http://purl.obolibrary.org/obo/DOID_9140
2115     37515  http://purl.obolibrary.org/obo/DOID_10138
2116       102  http://purl.obolibrary.org/obo/DOID_10371
2117       060   http://purl.obolibrary.org/obo/DOID_9682

[2118 rows x 2 columns]

Very nice, now we can do another simple left join on our diagnoses table to add our maps

In [7]:
diag_with_doid = diag_with_titles.merge(doid_icd_map, on="icd9_code", how="left")
print(str(len(diag_with_doid[diag_with_doid['doid'].notnull()])) + " out of " + str(len(diag_with_doid))) 

diag_with_doid

334 out of 1761


row_id  subject_id  hadm_id  seq_num icd9_code  \
0     112344       10006   142345        1     99591   
1     112345       10006   142345        2     99662   
2     112346       10006   142345        3      5672   
3     112347       10006   142345        4     40391   
4     112348       10006   142345        5     42731   
...      ...         ...      ...      ...       ...   
1756  397673       44228   103379        7      1975   
1757  397674       44228   103379        8     45182   
1758  397675       44228   103379        9     99592   
1759  397676       44228   103379       10      2449   
1760  397677       44228   103379       11      2859   

                                             long_title  \
0                                                Sepsis   
1     Infection and inflammatory reaction due to oth...   
2                                                   NaN   
3     Hypertensive chronic kidney disease, unspecifi...   
4                                   Atrial fibrillation   
...                                                 ...   
1756  Secondary malignant neoplasm of large intestin...   
1757  Phlebitis and thrombophlebitis of superficial ...   
1758                                      Severe sepsis   
1759                Unspecified acquired hypothyroidism   
1760                                Anemia, unspecified   

                                             doid  
0     http://purl.obolibrary.org/obo/DOID_0040085  
1                                             NaN  
2                                             NaN  
3                                             NaN  
4     http://purl.obolibrary.org/obo/DOID_0060224  
...                                           ...  
1756                                          NaN  
1757                                          NaN  
1758                                          NaN  
1759     http://purl.obolibrary.org/obo/DOID_1459  
1760     http://purl.obolibrary.org/obo/DOID_2355  

[1761 rows x 7 columns]

Depressingly, there were only 334 with a direct match! Thankfully we know that these things are matched.

In [8]:
from difflib import SequenceMatcher
match = list()
for index, row in diag_with_doid.iterrows():
    if pd.isnull(row['doid']) and row['doid'] not in match:
      match.append(row['icd9_code'])
     
ii = 0
for code in match:
    longest_match = False
    longest_doid = False
    for i, dent in doid_icd_map.iterrows():
        if len(dent['icd9_code']) < len(code):
            if code.find(dent['icd9_code']) == 0:
                if not longest_match or longest_match and len(longest_match) < len(dent['icd9_code']):
                    longest_match = dent['icd9_code']
                    longest_doid = dent['doid']
    if longest_match:
        print('ye')
        print(longest_doid)
        diag_with_doid.loc[diag_with_doid.icd9_code == code, 'doid'] = longest_doid
    
    ii+=1
    print(str(ii)+'/'+str(len(match)))

1/1427
2/1427
3/1427
ye
http://purl.obolibrary.org/obo/DOID_6000
4/1427
5/1427
6/1427
7/1427
8/1427
9/1427
10/1427
11/1427
ye
http://purl.obolibrary.org/obo/DOID_9351
12/1427
13/1427
14/1427
15/1427
ye
http://purl.obolibrary.org/obo/DOID_13189
16/1427
17/1427
18/1427
19/1427
20/1427
21/1427
22/1427
ye
http://purl.obolibrary.org/obo/DOID_10266
23/1427
24/1427
25/1427
26/1427
27/1427
28/1427
29/1427
30/1427
31/1427
32/1427
33/1427
ye
http://purl.obolibrary.org/obo/DOID_8516
34/1427
35/1427
36/1427
37/1427
ye
http://purl.obolibrary.org/obo/DOID_9351
38/1427
39/1427
40/1427
41/1427
ye
http://purl.obolibrary.org/obo/DOID_10825
42/1427
43/1427
44/1427
45/1427
46/1427
47/1427
48/1427
49/1427
ye
http://purl.obolibrary.org/obo/DOID_10825
50/1427
ye
http://purl.obolibrary.org/obo/DOID_2841
51/1427
52/1427
53/1427
54/1427
55/1427
56/1427
57/1427
ye
http://purl.obolibrary.org/obo/DOID_6000
58/1427
59/1427
ye
http://purl.obolibrary.org/obo/DOID_784
60/1427
61/1427
ye
http://purl.obolibrary.org/obo/

443/1427
444/1427
ye
http://purl.obolibrary.org/obo/DOID_10825
445/1427
ye
http://purl.obolibrary.org/obo/DOID_1724
446/1427
ye
http://purl.obolibrary.org/obo/DOID_10808
447/1427
448/1427
ye
http://purl.obolibrary.org/obo/DOID_10825
449/1427
450/1427
451/1427
452/1427
453/1427
454/1427
455/1427
456/1427
457/1427
458/1427
459/1427
460/1427
ye
http://purl.obolibrary.org/obo/DOID_9775
461/1427
462/1427
463/1427
464/1427
465/1427
ye
http://purl.obolibrary.org/obo/DOID_9351
466/1427
467/1427
ye
http://purl.obolibrary.org/obo/DOID_8717
468/1427
469/1427
470/1427
471/1427
ye
http://purl.obolibrary.org/obo/DOID_2089
472/1427
473/1427
ye
http://purl.obolibrary.org/obo/DOID_9775
474/1427
ye
http://purl.obolibrary.org/obo/DOID_6000
475/1427
476/1427
477/1427
ye
http://purl.obolibrary.org/obo/DOID_784
478/1427
479/1427
480/1427
481/1427
482/1427
ye
http://purl.obolibrary.org/obo/DOID_9351
483/1427
484/1427
ye
http://purl.obolibrary.org/obo/DOID_11476
485/1427
486/1427
487/1427
488/1427
489/1427
ye

848/1427
849/1427
850/1427
851/1427
852/1427
853/1427
854/1427
855/1427
ye
http://purl.obolibrary.org/obo/DOID_9743
856/1427
ye
http://purl.obolibrary.org/obo/DOID_10825
857/1427
ye
http://purl.obolibrary.org/obo/DOID_8717
858/1427
ye
http://purl.obolibrary.org/obo/DOID_8549
859/1427
860/1427
861/1427
862/1427
863/1427
864/1427
ye
http://purl.obolibrary.org/obo/DOID_8549
865/1427
866/1427
867/1427
ye
http://purl.obolibrary.org/obo/DOID_8717
868/1427
ye
http://purl.obolibrary.org/obo/DOID_9743
869/1427
870/1427
871/1427
872/1427
873/1427
874/1427
875/1427
ye
http://purl.obolibrary.org/obo/DOID_8549
876/1427
877/1427
878/1427
879/1427
880/1427
881/1427
ye
http://purl.obolibrary.org/obo/DOID_8717
882/1427
ye
http://purl.obolibrary.org/obo/DOID_8717
883/1427
884/1427
885/1427
886/1427
887/1427
ye
http://purl.obolibrary.org/obo/DOID_5200
888/1427
ye
http://purl.obolibrary.org/obo/DOID_9743
889/1427
ye
http://purl.obolibrary.org/obo/DOID_9500
890/1427
891/1427
892/1427
ye
http://purl.obolibr

1259/1427
1260/1427
1261/1427
ye
http://purl.obolibrary.org/obo/DOID_9351
1262/1427
1263/1427
1264/1427
1265/1427
1266/1427
1267/1427
1268/1427
1269/1427
1270/1427
1271/1427
1272/1427
1273/1427
1274/1427
ye
http://purl.obolibrary.org/obo/DOID_10825
1275/1427
1276/1427
1277/1427
1278/1427
1279/1427
ye
http://purl.obolibrary.org/obo/DOID_6000
1280/1427
1281/1427
1282/1427
1283/1427
ye
http://purl.obolibrary.org/obo/DOID_6000
1284/1427
1285/1427
1286/1427
1287/1427
1288/1427
1289/1427
ye
http://purl.obolibrary.org/obo/DOID_219
1290/1427
1291/1427
1292/1427
1293/1427
ye
http://purl.obolibrary.org/obo/DOID_784
1294/1427
ye
http://purl.obolibrary.org/obo/DOID_9351
1295/1427
1296/1427
1297/1427
1298/1427
1299/1427
ye
http://purl.obolibrary.org/obo/DOID_10247
1300/1427
1301/1427
1302/1427
1303/1427
1304/1427
1305/1427
ye
http://purl.obolibrary.org/obo/DOID_784
1306/1427
ye
http://purl.obolibrary.org/obo/DOID_9351
1307/1427
1308/1427
ye
http://purl.obolibrary.org/obo/DOID_9408
1309/1427
ye
http

In [9]:
print(str(len(diag_with_doid[diag_with_doid['doid'].notnull()])) + " out of " + str(len(diag_with_doid))) 

669 out of 1761


Now we've matched 

Bonus question: what's the time complexity of the above algorithm? How could it be improved?

In [10]:
diag_with_doid = diag_with_doid[diag_with_doid['doid'].notnull()]
diag_with_doid

row_id  subject_id  hadm_id  seq_num icd9_code  \
0     112344       10006   142345        1     99591   
4     112348       10006   142345        5     42731   
5     112349       10006   142345        6      4280   
6     112350       10006   142345        7      4241   
8     112352       10006   142345        9      2874   
...      ...         ...      ...      ...       ...   
1748  397652       44222   192189       10      3320   
1749  397653       44222   192189       11     72400   
1755  397672       44228   103379        6      1561   
1759  397676       44228   103379       10      2449   
1760  397677       44228   103379       11      2859   

                                         long_title  \
0                                            Sepsis   
4                               Atrial fibrillation   
5             Congestive heart failure, unspecified   
6                            Aortic valve disorders   
8                                               NaN   
...                                             ...   
1748                              Paralysis agitans   
1749            Spinal stenosis, unspecified region   
1755  Malignant neoplasm of extrahepatic bile ducts   
1759            Unspecified acquired hypothyroidism   
1760                            Anemia, unspecified   

                                             doid  
0     http://purl.obolibrary.org/obo/DOID_0040085  
4     http://purl.obolibrary.org/obo/DOID_0060224  
5        http://purl.obolibrary.org/obo/DOID_6000  
6          http://purl.obolibrary.org/obo/DOID_62  
8       http://purl.obolibrary.org/obo/DOID_11126  
...                                           ...  
1748    http://purl.obolibrary.org/obo/DOID_14330  
1749     http://purl.obolibrary.org/obo/DOID_6725  
1755     http://purl.obolibrary.org/obo/DOID_4606  
1759     http://purl.obolibrary.org/obo/DOID_1459  
1760     http://purl.obolibrary.org/obo/DOID_2355  

[669 rows x 7 columns]

You can see already from above that some are missing. sad. we could also be more advanced, and this is a whole field of ontology alignment that might allow us to do nice things here. For now we're just going to ignore everything that's not mapped.

So this can help us do things like data lookup. Let's say we're interested in hypertension, and different inds of hypertension

In [13]:
import csv
!komenti query -q "'heart disease'" -o DOID --class-mode --query-type subeq --out hd_classes.tsv
hd_classes = dict()
with open('hd_classes.tsv') as data:
  reader = csv.reader(data, delimiter='\t')
  for row in reader:
    if row[1] not in hd_classes:
        hd_classes[row[1]] = row[0]
hd_classes

SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Saved 702 labels for 255 terms to hd_classes.tsv


{'http://purl.obolibrary.org/obo/DOID_6419': 'tetralogy of fallot',
 'http://purl.obolibrary.org/obo/DOID_0110647': 'long qt syndrome 5',
 'http://purl.obolibrary.org/obo/DOID_0110648': 'long qt syndrome 6',
 'http://purl.obolibrary.org/obo/DOID_11516': 'hypertensive heart disease',
 'http://purl.obolibrary.org/obo/DOID_1882': 'atrial heart septal defect',
 'http://purl.obolibrary.org/obo/DOID_0060068': 'nonbacterial thrombotic endocarditis',
 'http://purl.obolibrary.org/obo/DOID_0110083': 'arrhythmogenic right ventricular dysplasia 12',
 'http://purl.obolibrary.org/obo/DOID_0110084': 'arrhythmogenic right ventricular dysplasia 13',
 'http://purl.obolibrary.org/obo/DOID_0110081': 'arrhythmogenic right ventricular dysplasia 10',
 'http://purl.obolibrary.org/obo/DOID_0110082': 'arrhythmogenic right ventricular dysplasia 11',
 'http://purl.obolibrary.org/obo/DOID_0110645': 'long qt syndrome 2',
 'http://purl.obolibrary.org/obo/DOID_0110646': 'long qt syndrome 3',
 'http://purl.obolibrary.

Now let's find some of these in our rows...

In [14]:
hd_admissions = diag_with_doid[diag_with_doid['doid'].isin(hd_classes.keys())]
hd_admissions['long_title'].value_counts()

Atrial fibrillation                                               48
Congestive heart failure, unspecified                             39
Aortic valve disorders                                             6
Chronic diastolic heart failure                                    6
Diseases of tricuspid valve                                        5
Acute systolic heart failure                                       4
Primary pulmonary hypertension                                     3
Cardiac arrest                                                     3
Other chronic pulmonary heart diseases                             3
Acute on chronic diastolic heart failure                           3
Other pulmonary embolism and infarction                            3
Acute diastolic heart failure                                      2
Rheumatic heart failure (congestive)                               2
Chronic systolic heart failure                                     2
Diastolic heart failure, unspecifi

Now try this with another disease or set of diseases you're interested in:

In [21]:
!komenti query -q "'heart disease'" -o DOID --class-mode --query-type subeq --out hd_classes.tsv
hd_classes = dict()
with open('hd_classes.tsv') as data:
  reader = csv.reader(data, delimiter='\t')
  for row in reader:
    if row[1] not in hd_classes:
        hd_classes[row[1]] = row[0]
print(hd_classes)
hd_admissions = diag_with_doid[diag_with_doid['doid'].isin(hd_classes.keys())]
hd_admissions

SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Saved 702 labels for 255 terms to hd_classes.tsv
{'http://purl.obolibrary.org/obo/DOID_6419': 'tetralogy of fallot', 'http://purl.obolibrary.org/obo/DOID_0110647': 'long qt syndrome 5', 'http://purl.obolibrary.org/obo/DOID_0110648': 'long qt syndrome 6', 'http://purl.obolibrary.org/obo/DOID_11516': 'hypertensive heart disease', 'http://purl.obolibrary.org/obo/DOID_1882': 'atrial heart septal defect', 'http://purl.obolibrary.org/obo/DOID_0060068': 'nonbacterial thrombotic endocarditis', 'http://purl.obolibrary.org/obo/DOID_0110083': 'arrhythmogenic right ventricular dysplasia 12', 'http://purl.obolibrary.org/obo/DOID_0110084': 'arrhythmogenic right ventricular dysplasia 13', 'http://purl.obolibrary.org/obo/DOID_0110081': 'arrhythmogenic right ventricular dysplasia 10', 'http://pur

row_id  subject_id  hadm_id  seq_num icd9_code  \
4     112348       10006   142345        5     42731   
5     112349       10006   142345        6      4280   
6     112350       10006   142345        7      4241   
31    112414       10013   165520        5     42731   
41    112449       10017   199207        6     41511   
...      ...         ...      ...      ...       ...   
1626  396243       43870   142633        9     42731   
1648  396310       43881   167021       12     41519   
1654  396316       43881   172454        1     41519   
1676  396373       43909   167612        6     42731   
1710  397346       44154   174245        5     42731   

                                        long_title  \
4                              Atrial fibrillation   
5            Congestive heart failure, unspecified   
6                           Aortic valve disorders   
31                             Atrial fibrillation   
41    Iatrogenic pulmonary embolism and infarction   
...                                            ...   
1626                           Atrial fibrillation   
1648       Other pulmonary embolism and infarction   
1654       Other pulmonary embolism and infarction   
1676                           Atrial fibrillation   
1710                           Atrial fibrillation   

                                             doid  
4     http://purl.obolibrary.org/obo/DOID_0060224  
5        http://purl.obolibrary.org/obo/DOID_6000  
6          http://purl.obolibrary.org/obo/DOID_62  
31    http://purl.obolibrary.org/obo/DOID_0060224  
41       http://purl.obolibrary.org/obo/DOID_8516  
...                                           ...  
1626  http://purl.obolibrary.org/obo/DOID_0060224  
1648     http://purl.obolibrary.org/obo/DOID_8516  
1654     http://purl.obolibrary.org/obo/DOID_8516  
1676  http://purl.obolibrary.org/obo/DOID_0060224  
1710  http://purl.obolibrary.org/obo/DOID_0060224  

[139 rows x 7 columns]

# Part 4: Semantic characterisation

So now we can start to think about how we can build phenotype profiles, and use the underlying semantic information the ontologies to start to discern between 

In [84]:
admissions = dict()
for index, row in diag_with_doid.iterrows():
    if row['hadm_id'] not in admissions:
        admissions[row['hadm_id']] = {
            'codes': list(),
            'contains_hd': False
        }
        
    if row['doid'] in hd_classes.keys():
        print('meow')
        admissions[row['hadm_id']]['contains_hd'] = True
    #else:
    admissions[row['hadm_id']]['codes'].append(row['doid'])

a = 0
with open('klar_input.tsv', 'w') as f:
    for aid, v in admissions.items():
      a+= 1
      print(a)
      groups = list()
      if v['contains_hd']:
        groups.append('HD')
        print('meow2')
      else:
        groups.append('NoHD')
    
      if len(groups) != 0:
        print('meow3')
        line = "\t".join((str(aid), ";".join(v['codes']), ";".join(groups)))
        print(line)
        f.write(line + '\n')
        

meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
meow
1
meow2
meow3
142345	http://purl.obolibrary.org/obo/DOID_0040085;http://purl.obolibrary.org/obo/DOID_0060224;http://purl.obolibrary.org/obo/DOID_6000;http://purl.obolibrary.org/obo/DOID_62;http://purl.obolibrary.org/obo/DOID_11126;http://purl.obolibrary.org/obo/DOID_9351;http://purl.obolibrary.org/obo/DO

In [85]:
!cat klar_input.tsv

142345	http://purl.obolibrary.org/obo/DOID_0040085;http://purl.obolibrary.org/obo/DOID_0060224;http://purl.obolibrary.org/obo/DOID_6000;http://purl.obolibrary.org/obo/DOID_62;http://purl.obolibrary.org/obo/DOID_11126;http://purl.obolibrary.org/obo/DOID_9351;http://purl.obolibrary.org/obo/DOID_13189	HD
105331	http://purl.obolibrary.org/obo/DOID_2237;http://purl.obolibrary.org/obo/DOID_1588	NoHD
165520	http://purl.obolibrary.org/obo/DOID_10266;http://purl.obolibrary.org/obo/DOID_0060224	HD
199207	http://purl.obolibrary.org/obo/DOID_9675;http://purl.obolibrary.org/obo/DOID_8516;http://purl.obolibrary.org/obo/DOID_9351;http://purl.obolibrary.org/obo/DOID_10825	HD
177759	http://purl.obolibrary.org/obo/DOID_11162;http://purl.obolibrary.org/obo/DOID_2913;http://purl.obolibrary.org/obo/DOID_13413;http://purl.obolibrary.org/obo/DOID_12351;http://purl.obolibrary.org/obo/DOID_10825;http://purl.obolibrary.org/obo/DOID_2841	NoHD
103770	http://purl.obolibrary.org/obo/DOID_10823;http://purl.obolibrar

In [87]:
!klarigi/bin/klarigi --data klar_input.tsv --ontology doid.owl --power

HD: Scoring completed. Candidates: 335
NoHD: Scoring completed. Candidates: 335
----------------
Group: HD (78 members)
Overall inclusion: 98.72%
Overall exclusion: 97.87%
Explanatory classes:
  IRI: atrial fibrillation (http://purl.obolibrary.org/obo/DOID_0060224), Power: 0.62 (inc: 0.62, exc: 1.0), IC: 0.8
  IRI: congestive heart failure (http://purl.obolibrary.org/obo/DOID_6000), Power: 0.63 (inc: 0.63, exc: 1.0), IC: 0.67
----------------

----------------
Group: NoHD (47 members)
Overall inclusion: 61.7%
Overall exclusion: 98.72%
Explanatory classes:
  IRI: respiratory failure (http://purl.obolibrary.org/obo/DOID_11162), Power: 0.15 (inc: 0.34, exc: 0.81), IC: 0.75
  IRI: hepatobiliary disease (http://purl.obolibrary.org/obo/DOID_3118), Power: 0.18 (inc: 0.23, exc: 0.95), IC: 0.49
  IRI: liver disease (http://purl.obolibrary.org/obo/DOID_409), Power: 0.17 (inc: 0.21, exc: 0.96), IC: 0.58
  IRI: hematologic cancer (http://purl.obolibrary.org/obo/DOID_2531), Power: 0.13 (inc: 0.13, 

# Part 5: Digital Phenotyping

 So from this we can see a little of the non heart and heart disease diseases, and what characterise them. in particular we can see that brain diseases are infrequently co-morbid with heart diseases. 
 
#Part 3: Digital PhenotypingPart

Now it's time to 

In [37]:
labevents = pd.read_csv("mimic/LABEVENTS.csv")
labdefs = pd.read_csv("mimic/D_LABITEMS.csv")[['itemid','label']]

In [41]:
labevents = labevents.merge(labdefs, on="itemid", how="left")
labevents = labevents[labevents['hadm_id'].notnull()]

labevents

row_id  subject_id   hadm_id  itemid            charttime value  \
0       6245198       10006  142345.0   50813  2164-10-23 17:33:00   4.4   
1       6245199       10006  142345.0   50861  2164-10-23 17:38:00     9   
2       6245200       10006  142345.0   50862  2164-10-23 17:38:00   3.4   
3       6245201       10006  142345.0   50863  2164-10-23 17:38:00   167   
4       6245202       10006  142345.0   50868  2164-10-23 17:38:00    20   
...         ...         ...       ...     ...                  ...   ...   
61807  20452679       44228  103379.0   51250  2170-12-24 04:09:00    88   
61808  20452680       44228  103379.0   51265  2170-12-24 04:09:00   595   
61809  20452681       44228  103379.0   51277  2170-12-24 04:09:00  14.5   
61810  20452682       44228  103379.0   51279  2170-12-24 04:09:00  2.76   
61811  20452683       44228  103379.0   51301  2170-12-24 04:09:00  16.9   

       valuenum valueuom      flag                           label  
0          4.40   mmol/L  abnormal                         Lactate  
1          9.00     IU/L       NaN  Alanine Aminotransferase (ALT)  
2          3.40     g/dL       NaN                         Albumin  
3        167.00     IU/L  abnormal            Alkaline Phosphatase  
4         20.00    mEq/L       NaN                       Anion Gap  
...         ...      ...       ...                             ...  
61807     88.00       fL       NaN                             MCV  
61808    595.00     K/uL  abnormal                  Platelet Count  
61809     14.50        %       NaN                             RDW  
61810      2.76     m/uL  abnormal                 Red Blood Cells  
61811     16.90     K/uL  abnormal               White Blood Cells  

[61812 rows x 10 columns]

So we have now all of these nice blood measurements, but how do we apply them to anything?

In [42]:
blood_ranges = pd.read_csv('blood_ranges.tsv', sep='\t')
blood_ranges

itemid            label   lower  upper increased_phen decreased_phen
0   50983     blood sodium  135.00  145.0     HP:0003228     HP:0002902
1   50809    blood glucose   79.00  110.0     HP:0003074     HP:0001943
2   50971  blood potassium    3.70    5.2     HP:0002153     HP:0002900
3   50902   blood chloride   98.00  106.0     HP:0011423     HP:0003113
4   51002       troponin i    0.30    NaN     HP:0410173            NaN
5   51003       troponin t    0.02    NaN     HP:0410173            NaN

So what we want to do is find people who have test scores outside of this 

In [61]:
admission_phenotypes = dict()
for i, row in labevents.iterrows():
  if len(diag_with_doid[diag_with_doid['hadm_id'] == row['hadm_id']]) != 0:
    if row['hadm_id'] not in admission_phenotypes:
      admission_phenotypes[row['hadm_id']] = list()
    for z, ra in blood_ranges.iterrows():
      if row['itemid'] == ra['itemid']:
        try:
          value = float(row['value'].replace('<','').replace('>',''))
        except:
          continue
        if value < float(ra['lower']) and pd.notnull(ra['decreased_phen']) and ra['decreased_phen'] not in admission_phenotypes[row['hadm_id']]:
          admission_phenotypes[row['hadm_id']].append(ra['decreased_phen'])
        elif value > float(ra['upper']) and pd.notnull(ra['increased_phen']) and ra['increased_phen'] not in admission_phenotypes[row['hadm_id']]:
          admission_phenotypes[row['hadm_id']].append(ra['increased_phen'])
        
admission_phenotypes

{142345.0: ['HP:0003113', 'HP:0002153'],
 105331.0: ['HP:0011423', 'HP:0002153', 'HP:0002900', 'HP:0003228'],
 165520.0: [],
 199207.0: ['HP:0002900'],
 177759.0: ['HP:0002900', 'HP:0003074', 'HP:0003113', 'HP:0002153'],
 103770.0: ['HP:0003113', 'HP:0002902', 'HP:0002900'],
 199395.0: ['HP:0003074',
  'HP:0011423',
  'HP:0001943',
  'HP:0002153',
  'HP:0002900',
  'HP:0003228'],
 132349.0: ['HP:0011423', 'HP:0002900', 'HP:0003228'],
 140372.0: ['HP:0011423'],
 157235.0: ['HP:0002153', 'HP:0011423'],
 110244.0: [],
 189483.0: ['HP:0011423', 'HP:0003228', 'HP:0002900'],
 111115.0: ['HP:0011423', 'HP:0002900', 'HP:0003228'],
 157839.0: ['HP:0011423', 'HP:0002900'],
 148562.0: ['HP:0002900',
  'HP:0011423',
  'HP:0003074',
  'HP:0002902',
  'HP:0002153'],
 168674.0: ['HP:0011423', 'HP:0003228'],
 124073.0: ['HP:0002900', 'HP:0002902', 'HP:0011423'],
 126949.0: ['HP:0003113',
  'HP:0002902',
  'HP:0003074',
  'HP:0011423',
  'HP:0002900'],
 100375.0: ['HP:0003113', 'HP:0002902', 'HP:000290

In [93]:
a = 0
with open('klar_input.tsv', 'w') as f:
    for aid, v in admissions.items():
      a+= 1
      print(a)
      groups = list()
      if v['contains_hd']:
        groups.append('HD')
      else:
        groups.append('NoHD')
    
      if len(groups) != 0 and len(admission_phenotypes[aid]) > 0:
        line = "\t".join((str(aid), ";".join(admission_phenotypes[aid]), ";".join(groups)))
        print(line)
        f.write(line + '\n')

1
142345	HP:0003113;HP:0002153	HD
2
105331	HP:0011423;HP:0002153;HP:0002900;HP:0003228	NoHD
3
4
199207	HP:0002900	HD
5
177759	HP:0002900;HP:0003074;HP:0003113;HP:0002153	NoHD
6
103770	HP:0003113;HP:0002902;HP:0002900	NoHD
7
199395	HP:0003074;HP:0011423;HP:0001943;HP:0002153;HP:0002900;HP:0003228	HD
8
132349	HP:0011423;HP:0002900;HP:0003228	HD
9
140372	HP:0011423	HD
10
157235	HP:0002153;HP:0011423	HD
11
12
189483	HP:0011423;HP:0003228;HP:0002900	NoHD
13
111115	HP:0011423;HP:0002900;HP:0003228	HD
14
157839	HP:0011423;HP:0002900	HD
15
148562	HP:0002900;HP:0011423;HP:0003074;HP:0002902;HP:0002153	HD
16
168674	HP:0011423;HP:0003228	HD
17
124073	HP:0002900;HP:0002902;HP:0011423	HD
18
126949	HP:0003113;HP:0002902;HP:0003074;HP:0011423;HP:0002900	HD
19
100375	HP:0003113;HP:0002902;HP:0002900	HD
20
122098	HP:0011423;HP:0002153;HP:0002902;HP:0003113;HP:0002900;HP:0001943;HP:0003074	NoHD
21
142582	HP:0011423;HP:0002153;HP:0003228	NoHD
22
145203	HP:0002902;HP:0011423;HP:0002900;HP:0003113;HP:00032

In [58]:
admissions

{142345: {'codes': ['http://purl.obolibrary.org/obo/DOID_0040085',
   'http://purl.obolibrary.org/obo/DOID_0060224',
   'http://purl.obolibrary.org/obo/DOID_6000',
   'http://purl.obolibrary.org/obo/DOID_62',
   'http://purl.obolibrary.org/obo/DOID_11126',
   'http://purl.obolibrary.org/obo/DOID_9351',
   'http://purl.obolibrary.org/obo/DOID_13189'],
  'contains_hd': True},
 105331: {'codes': ['http://purl.obolibrary.org/obo/DOID_2237',
   'http://purl.obolibrary.org/obo/DOID_1588'],
  'contains_hd': False},
 165520: {'codes': ['http://purl.obolibrary.org/obo/DOID_10266',
   'http://purl.obolibrary.org/obo/DOID_0060224'],
  'contains_hd': True},
 199207: {'codes': ['http://purl.obolibrary.org/obo/DOID_9675',
   'http://purl.obolibrary.org/obo/DOID_8516',
   'http://purl.obolibrary.org/obo/DOID_9351',
   'http://purl.obolibrary.org/obo/DOID_10825'],
  'contains_hd': True},
 177759: {'codes': ['http://purl.obolibrary.org/obo/DOID_11162',
   'http://purl.obolibrary.org/obo/DOID_2913',
   

In [94]:
!klarigi/bin/klarigi --data klar_input.tsv --ontology hp.owl

HD: Scoring completed. Candidates: 26
NoHD: Scoring completed. Candidates: 26
----------------
Group: HD (75 members)
Overall inclusion: 85.33%
Overall exclusion: 97.73%
Explanatory classes:
  IRI: Hypernatremia (http://purl.obolibrary.org/obo/HP_0003228), Power: 0.16 (inc: 0.41, exc: 0.75), IC: 0.81
  IRI: Abnormal blood oxygen level (http://purl.obolibrary.org/obo/HP_0500165), Power: 0.01 (inc: 0.37, exc: 0.64), IC: 0.67
  IRI: Abnormal blood sodium concentration (http://purl.obolibrary.org/obo/HP_0010931), Power: 0.09 (inc: 0.71, exc: 0.39), IC: 0.74
----------------

----------------
Group: NoHD (44 members)
Overall inclusion: 70.45%
Overall exclusion: 98.67%
Explanatory classes:
  IRI: Hyponatremia (http://purl.obolibrary.org/obo/HP_0002902), Power: 0.14 (inc: 0.5, exc: 0.64), IC: 0.81
  IRI: Hypochloremia (http://purl.obolibrary.org/obo/HP_0003113), Power: 0.0 (inc: 0.36, exc: 0.64), IC: 0.77
  IRI: Hyperglycemia (http://purl.obolibrary.org/obo/HP_0003074), Power: 0.0 (inc: 0.36,

In [98]:
diag_with_titles[['hadm_id','long_title']]

hadm_id                                         long_title
0      142345                                             Sepsis
1      142345  Infection and inflammatory reaction due to oth...
2      142345                                                NaN
3      142345  Hypertensive chronic kidney disease, unspecifi...
4      142345                                Atrial fibrillation
...       ...                                                ...
1756   103379  Secondary malignant neoplasm of large intestin...
1757   103379  Phlebitis and thrombophlebitis of superficial ...
1758   103379                                      Severe sepsis
1759   103379                Unspecified acquired hypothyroidism
1760   103379                                Anemia, unspecified

[1761 rows x 2 columns]

In [101]:

labdefs = pd.read_csv("mimic/D_LABITEMS.csv")
labdefs[labdefs['itemid'] == 50813]


row_id  itemid    label  fluid   category loinc_code
13      14   50813  Lactate  Blood  Blood Gas    32693-4